Consider value objects (#54)
andycroll/andycroll.com: _posts/2019/2019-07-28-consider-value-objects.md

---
title: "Consider Value Objects"
description: "Making text readable on"
layout: article
category: ruby
image:
  base: '2019/consider-value-objects'
  alt: 'Cubes'
  credit: Christian Fregnan
  source: "https://unsplash.com/photos/TAVKURx-xLw"

---

> In your applications, you’ll often find yourself creating multiple view helpers around the same concept, constructing complex calculation methods, or passing a value (or several) around multiple methods.

Reviewer: * a range of methods  (since you already use 'multiple')

Reviewer: I mean this isn't a big deal...

> In your applications, you’ll often find yourself creating multiple view helpers around the same concept, constructing complex calculation methods, or passing a value (or several) around multiple methods.

Author: ```suggestion
In your applications, you’ll often find yourself creating a range of view helpers around the same concept, constructing complex calculation methods, or passing a value (or several) around multiple methods.
```

> When this happens it is possible a simple object, smaller than an Active Record model, is trying to reveal itself.

Reviewer: Comma needed after 'happens'

> When this happens it is possible a simple object, smaller than an Active Record model, is trying to reveal itself.

Author: ```suggestion
When this happens a simple object, smaller than an Active Record model, is trying to reveal itself.
```

> A small simple object, like money or a date range, whose equality isn’t based on identity.

## Instead of...

...just using a helper:

```ruby
def css_color(r, g, b)
  "##{r.to_s(16)}#{g.to_s(16)}#{b.to_s(16)}"
end
```

## Use...

...an extracted “Value Object” to truly capture the concept your application is using.

```ruby
class RGBColor
  def initialize(red, green, blue)
    @red = [[0, red].max], 255].min
    @green = [[0, green].max], 255].min
    @blue = [[0, blue].max], 255].min      
  end

  def to_hex
    [@red, @green, @blue].each_with_object("") do |part, to_hex|
      to_hex << part.to_s(16)
    end
  end
end
```

### And in the helper

```ruby
def css_color(r, g, b)
  "##{RGBColor.new(r, g, b).to_hex}"
end
```

## Why?

Just because a concept in our app isn't saved to the database doesn't mean it shouldn’t be an object in our app.

Refactoring into a value object has several benefits since you isolate the code for this concept. You might hear this referred to as ‘separation of concerns’. You can also test the behaviour of your concept in isolation, leading to more comprehensive and efficient tests.

> This improved organisation of your code will clarify your understanding. The code/concept is also easier to reuse and you gain a clear place for extending the functionality of the object.

Reviewer: Yes — all flows much better.

---
title: "Consider Value Objects"
description: "Making text readable on"
layout: article
category: ruby
image:
  base: '2019/consider-value-objects'
  alt: 'Cubes'
  credit: Christian Fregnan
  source: "https://unsplash.com/photos/TAVKURx-xLw"

---

> You’ll often find yourself doing creating multiple view helpers around the same concept, creating complex calculation methods, or passing a value (or several) around multiple methods.

Reviewer: ```suggestion
In your applications, you’ll often find yourself creating multiple view helpers around the same concept, constructing complex calculation methods, or passing a value (or several) around multiple methods.
```

When this happens it is possible a simple object, smaller than an Active Record model, is trying to reveal itself.

In these cases, consider refactoring the functionality into a “Value Object”.

## Instead of...

...just using a helper:

```ruby
def css_color(r, g, b)
  "##{r.to_s(16)}#{g.to_s(16)}#{b.to_s(16)}"
end
```

## Use...

...an extracted “Value Object” to truly capture the concept your application is using.

```ruby
class RGBColor
  def initialize(red, green, blue)
    @red = [[0, red].max], 255].min
    @green = [[0, green].max], 255].min
    @blue = [[0, blue].max], 255].min      
  end

  def to_hex
    [@red, @green, @blue].each_with_object("") do |part, to_hex|
      to_hex << part.to_s(16)
    end
  end
end
```

### And in the helper

```ruby
def css_color(r, g, b)
  "##{RGBColor.new(r, g, b).to_hex}"
end
```

> ## Why?

Reviewer: This section reads like a series of bullet points. It seems jumpy. The sentences don't seem to flow onto one another.

Just because a concept in our app isn't saved to the database doesn't mean it shouldn’t be an object in our app.

Martin Fowler [defines a Value Object](https://martinfowler.com/eaaCatalog/valueObject.html) as:

> A small simple object, like money or a date range, whose equality isn’t based on identity.

> Refactoring into a value object has several benefits as you isolate the code for this concept. You might hear this referred to as ‘separation of concerns’.

Reviewer: *'_since_ you isolate' or '_because_ you isolate'...dunno why, but 'as' doesn't feel so right.

Your can test the behaviour of your concept in isolation, leading to more comprehensive and efficient tests.

This improved organisation of your code will clarify your understanding. The code is also easier to reuse and you have a clear place for extending the functionality of the object.

## Why not?

Finding the correct time to extract a concept into a value object can be difficult. Too early leads to unnecessary complication and potentially increased confusion; too late and you’re already coding in a mess!

> 

Reviewer: Should this part be in the 'Why'? Since it's showing a way to implement it despite saying avoid `Struct` and `OpenStruct`?

I'm torn!

Author: What if it opens with "It might be too soon. Finding..."

Reviewer: I was actually talking about the sentence afterwards, but this addition works really well.

> For implementation you could use a `Struct` or even `OpenStruct` rather than the class definition approach shown in the example, but these are mutable objects (you can change their properties) which can add complication rather than to simplify.

Reviewer: *..can add complication, rather than simplify your code.

---
title: "Consider Value Objects"
description: "Making text readable on"
layout: article
category: ruby
image:
  base: '2019/consider-value-objects'
  alt: 'Cubes'
  credit: Christian Fregnan
  source: "https://unsplash.com/photos/TAVKURx-xLw"

---

Often we use simple concepts in our app, that don't start out as objects.

You'll often find yourself creating multiple view helpers or passing a value (or several) around several methods, or within a model.

> This loose concept can be refactored into a “Value Object”.

Reviewer: I think these three opening sentences need re-working. I was confused.

Maybe something like:

- You'll often find yourself doing x, y, z...
- This happens in cases where you want something simpler than a fully-blown objecct
- In these cases, consider refactoring into a 'value object'.

A structure like that. What do you think?

## Instead of...

> ...just using a helper.

Reviewer: And you didn't refer to helpers in your intro, maybe you should.

Also, colon, not full-stop. :-D

```ruby
def css_color(r, g, b)
  "##{r.to_s(16)}#{g.to_s(16)}#{b.to_s(16)}"
end
```

## Use...

...an extracted “Value Object” to truly capture the concept your application is using.

```ruby
class RGBColor
  def initialize(red, green, blue)
    @red = [[0, red].max], 255].min
    @green = [[0, green].max], 255].min
    @blue = [[0, blue].max], 255].min      
  end

>   #

Reviewer: Meant to be commented out?

  # def self.from_hex(hex)
  #   new(hex[0, 2], hex[2, 2], hex[4, 2])
  # end

  def to_hex
    [@red, @green, @blue].each_with_object("") do |part, to_hex|
      to_hex << part.to_s(16)
    end
  end
end
```

### And in the helper

```ruby
def css_color(r, g, b)
  "##{RGBColor.new(r, g, b).to_hex}"
end
```

## Why?

Just because a concept in our app isn't saved to the database doesn't mean it shouldn’t be an object in our app.

Martin Fowler [defines a Value Object](https://martinfowler.com/eaaCatalog/valueObject.html) as:

> A small simple object, like money or a date range, whose equality isn’t based on identity.

> Refactoring into an object has several benefits as you isolate and separate the code for this concept. You might hear this referred to as ‘separation of concerns’.

Reviewer: ```suggestion
Refactoring into a value object has several benefits as you isolate the code for this concept. You might hear this referred to as ‘separation of concerns’.
```

> Your can test the behaviour of your concept in isolation. Thus more comprehensively and efficiently.

Reviewer: ```suggestion
Your can test the behaviour of your concept separately (or if you keep 'separate' above, then keep isolation here), leading to more comprehensive and efficient tests.
```

> The improved understanding and organisation of your code improves your understanding and you gain the benefits of reuse as well as a place for extensions to the concept you've created.

Reviewer: Sentence is a bit long and muddly.

Reviewer: You've got 'improved understanding' and 'improves your understanding' in the same sentence.  

Maybe just start with 'improved organisation'...'helps your understanding'. Then: the code is also easier to reuse and you have a clear place for extending the functionality of the object/the 'helpers'.

## Why not?

> It might be too complex, too soon.

Reviewer: I would take this sentence out and go straight into the next one.

> Finding the correct time to extract a concept can be difficult. Too early leads to over-complication and confusion, too late and you’re already coding in a mess!

Reviewer: *to extract a concept into a value object

Reviewer: ```suggestion
Finding the correct time to extract a concept into a value object can be difficult. Too early leads to unnecessary complication and potentially increased confusion; too late and you’re already coding in a mess!
```

> For implementation you could use a `Struct` or even `OpenStruct` rather than the class definition approach shown in the example, but they are mutable objects (you can change their values) which can add complication rather than help to simplify.

Reviewer: *but these are

And when you say 'change their values', do you mean their properties?

*rather than simplify
